dev/cluster.go

```diff
@@ -0,0 +1,288 @@
+	"bytes"
+	"context"
+	goerrors "errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+
+	"github.com/go-logr/logr"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var defaultSchemeBuilder runtime.SchemeBuilder = runtime.SchemeBuilder{
+	clientgoscheme.AddToScheme,
+	apiextensionsv1.AddToScheme,
+}
+
+type ClusterConfig struct {
+	Logger        logr.Logger
+	SchemeBuilder runtime.SchemeBuilder
+	NewWaiter     ClusterNewWaiterFunc
+	WaitOptions   []WaitOption
+}
+
+type ClusterNewWaiterFunc func(
+	client client.Client, scheme *runtime.Scheme,
+	defaultOpts ...WaitOption,
+) *Waiter
+
+func (c *ClusterConfig) Default() {
+	if c.Logger.GetSink() == nil {
+		c.Logger = logr.Discard()
+	}
+	if c.NewWaiter == nil {
+		c.NewWaiter = NewWaiter
+	}
+	if c.WaitOptions == nil {
+		c.WaitOptions = append(c.WaitOptions, WithLogger(c.Logger))
+	}
+}
+
+type ClusterOption interface {
+	ApplyToClusterConfig(c *ClusterConfig)
+}
+
+// Container object to hold kubernetes client interfaces and configuration.
+type Cluster struct {
+	Scheme     *runtime.Scheme
+	RestConfig *rest.Config
+	CtrlClient client.Client
+	Waiter     *Waiter
+	WorkDir    string
+	// Path to the kubeconfig of the cluster
+	Kubeconfig string
+
+	config ClusterConfig
+}
+
+// Creates a new Cluster object to interact with a Kubernetes cluster.
+func NewCluster(workDir string, opts ...ClusterOption) (*Cluster, error) {
+	c := &Cluster{
+		Scheme:     runtime.NewScheme(),
+		WorkDir:    workDir,
+		Kubeconfig: workDir + "/kubeconfig.yaml",
+	}
+
+	// Add default schemes
+	if err := defaultSchemeBuilder.AddToScheme(c.Scheme); err != nil {
+		return nil, fmt.Errorf("adding defaults to scheme: %w", err)
+	}
+
+	// Apply Options
+	for _, opt := range opts {
+		opt.ApplyToClusterConfig(&c.config)
+	}
+	// Apply schemes from Options
+	if c.config.SchemeBuilder != nil {
+		if err := c.config.SchemeBuilder.AddToScheme(c.Scheme); err != nil {
+			return nil, fmt.Errorf("adding to scheme: %w", err)
+		}
+	}
+
+	var err error
+	// Create RestConfig
+	c.RestConfig, err = clientcmd.BuildConfigFromFlags("", c.Kubeconfig)
+	if err != nil {
+		return nil, fmt.Errorf("getting rest.Config from kubeconfig: %w", err)
+	}
+
+	// Create Controller Runtime Client
+	c.CtrlClient, err = client.New(c.RestConfig, client.Options{
+		Scheme: c.Scheme,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("creating new ctrl client: %w", err)
+	}
+
+	c.Waiter = c.config.NewWaiter(c.CtrlClient, c.Scheme, c.config.WaitOptions...)
+
+	return c, nil
+}
+
+// Load kube objects from a list of http urls,
+// create these objects and wait for them to be ready.
+func (c *Cluster) CreateAndWaitFromHttp(
```

Comment:
In what cases would we prefer this over `CreateAndWaitFromFiles` ? Seems we are adding the potential to fail on HTTP get. Personally I would always add whatever manifests is required to the codebase itself and avoid the HTTP call alltogether.

Reply:
Does not hurt to keep it around though !